src/essence/Tools/Isochrone/IsochroneTool.js

```diff
@@ -0,0 +1,539 @@
+                } else {
+                    img.data[di] = 0;
+                    img.data[di + 1] = 0;
+                    img.data[di + 2] = 0;
+                    img.data[di + 3] = 0;
+                }
+                di += 4;
+            }
+        }
+        ctx.putImageData(img, 0, 0);
+        return tile;
+    }
+});
+
+function backlinkToMove(link) {
+    return [
+        [0, 1],
+        [1, 2],
+        [1, 1],
+        [2, 1],
+        [1, 0],
+        [2, -1],
+        [1, -1],
+        [1, -2],
+        [0, -1],
+        [-1, -2],
+        [-1, -1],
+        [-2, -1],
+        [-1, 0],
+        [-2, 1],
+        [-1, 1],
+        [-1, 2]
+    ][link - 1];
+}
+
+/////////////////////////////////////////////////
+
+const IsochroneTool = {
+    height: 0,
+    width: 250,
+    MMGISInterface: null,
+    vars: null,
+    dataSources: {},
+    containerEl: null,
+
+    managers: [],
+    activeManager: null,
+    managerCounter: 0,
+
+    hovered: false,
+    lastHoverCall: 0,
+    hoverPolyline: null,
+
+    colorRamps: [
+        [ //Red 5
+            [254, 229, 217],
+            [252, 187, 161],
+            [252, 106, 74],
+            [222, 45, 38],
+            [165, 15, 21]
+        ],
+        [ //Orange 4
+            [255, 255, 178],
+            [254, 204, 92],
+            [253, 141, 60],
+            [240, 59, 32],
+            [189, 0, 38]
+        ],
+        [ //Green 1
+            [237, 248, 233],
+            [186, 228, 179],
+            [116, 196, 118],
+            [49, 163, 84],
+            [0, 109, 44]
+        ],
+        [ //Blue 4
+            [240, 249, 232],
+            [186, 228, 188],
+            [123, 204, 196],
+            [67, 162, 202],
+            [8, 104, 172]
+        ],
+        [ //Purple 4
+            [237, 248, 251],
+            [179, 205, 227],
+            [140, 150, 198],
+            [136, 86, 167],
+            [129, 15, 124]
+        ]
+    ],
+
+    initialize: function() {
+        this.vars = L_.getToolVars('isochrone');
+
+        //add info for scaling low-res tiles
+        for(const dataType in this.vars.data) {
+            this.dataSources[dataType] = [];
+            for(const src of this.vars.data[dataType]) {
+                const {name, tileurl, minZoom, maxNativeZoom, resolution} = src;
+                let zoomOffset = 0;
+                switch(resolution) {
+                    case 256:
+                        zoomOffset = 0;
+                    break;
+                    case 128:
+                        zoomOffset = 1;
+                    break;
+                    case 64:
+                        zoomOffset = 2;
+                    break;
+                    case 32:
+                        zoomOffset = 3;
+                    break;
+                    default:
+                        console.warn(`IsochroneTool: ${dataType} source "${src.name}" has undefined or nonstandard resolution: ${src.resolution}!`);
```

Comment:
what would the user see in this case? does the layer still render? if not, is there a way to guarantee the layer always shows up despite the zoom level?

Reply:
Behavior now is that this data source would be removed as an option. If all data sources were removed and a model needed that kind of data then that would be caught by the new error handling.